parsl/providers/local/local.py

```diff
@@ -76,19 +76,70 @@ def status(self, job_ids):
 
         logger.debug("Checking status of: {0}".format(job_ids))
         for job_id in self.resources:
-
-            retcode, stdout, stderr = self.channel.execute_wait('ps -p {} > /dev/null 2> /dev/null; echo "STATUS:$?" '.format(
-                self.resources[job_id]['remote_pid']), self.cmd_timeout)
-            for line in stdout.split('\n'):
-                if line.startswith("STATUS:"):
-                    status = line.split("STATUS:")[1].strip()
-                    if status == "0":
-                        self.resources[job_id]['status'] = JobStatus(JobState.RUNNING)
+            # This job dict should really be a class on its own
+            job_dict = self.resources[job_id]
+            if job_dict['status'] is not None and job_dict['status'].terminal:
+                # We already checked this and it can't change after that
+                continue
+            # Script path should point to remote path if _should_move_files() is True
+            script_path = job_dict['script_path']
+
+            alive = self._is_alive(job_dict)
+            str_ec = self._read_job_file(script_path, '.ec').strip()
+
+            status = None
+            if str_ec == '-':
+                if alive:
+                    status = JobStatus(JobState.RUNNING)
+                else:
+                    # not alive but didn't get to write an exit code
+                    if 'cancelled' in job_dict:
+                        # because we cancelled it
+                        status = JobStatus(JobState.CANCELLED)
+                    else:
+                        # we didn't cancel it, so it must have been killed by something outside
+                        # parsl; we don't have a state for this, but we'll use CANCELLED with
+                        # a specific message
```

Comment:
I think this case would also be hit if something failed to start really early on? I've encountered a case where an unexpected shell was being used, for example. I don't think that needs a change if so, other than perhaps in the comment, but checking my understanding of the logic.

Reply:
Good catch. I was under the impression that this will be routed through /bin/bash, but I cannot see why that would be so. I think this should probably be changed to /bin/bash -c ...

Reply:
I wrapped this in /bin/bash -c. That said, I also tried with /bin/dash -c, and it worked. I looked at it and it didn't seem to contain any bashisms. That said, it is hard to imagine a scenario, short of bash not being installed, in which being explicit about the shell is a bad thing.